layers.go

```diff
@@ -2567,14 +2566,22 @@ func (r *layerStore) ApplyDiffWithDiffer(to string, options *drivers.ApplyDiffWi
 			},
 		}
 	}
-	output, err := ddriver.ApplyDiffWithDiffer(layer.ID, layer.Parent, options, differ)
+	out, err := ddriver.ApplyDiffWithDiffer(layer.ID, layer.Parent, options, differ)
```

Comment:
⚠️ 

I don’t think this works for pre-existing layers (`layer != nil`); if the layer is unlocked:
- Two callers can get here simultaneously, and write over each others’ `getStagingDir`
- The layer can be deleted concurrently to this operation!

---

c/image _only ever_ calls `ApplyDiffWithDiffer` without the layer ID, that’s the only code path I _fully_ care about. (The other code path is used in `cmd/containers-storage` and in tests, per comments in `store.go`.)

 It is very tempting to me to have a new specialized `PrepareStagedLayer` (to match the names of `{Apply,Cleanup}StagedLayer`) that only supports this simpler care; then several of the operations here could just disappear.

But `ApplyDiffWithDiffer` is now a stable API; I don’t immediately know how to deal with that. The concurrent layer deletion _might_ be possible to handle by holding a read lock on the layer store only (assuming all operations in `pkg/chunked` lock the layer store _only_ for reading, **and that’s not currently the case**), but the exclusion of two concurrent callers, and of any possible concurrent readers, is fundamentally the existing exclusive layer store lock.

Outright refusing to do the operation on existing layers … is code that could be written… but probably shouldn’t be merged. I don’t know.

Reply:
> But `ApplyDiffWithDiffer` is now a stable API; I don’t immediately know how to deal with that. The concurrent layer deletion _might_ be possible to handle by holding a read lock on the layer store only (assuming all operations in `pkg/chunked` lock the layer store _only_ for reading, **and that’s not currently the case**), but the exclusion of two concurrent callers, and of any possible concurrent readers, is fundamentally the existing exclusive layer store lock.
> 
> Outright refusing to do the operation on existing layers … is code that could be written… but probably shouldn’t be merged. I don’t know.

I don't know either how to solve it, unless introducing another per-layer lock file.  As you said, this code path is used only by `cmd/containers-storage`.

Could we just document this behavior, and warn any potential user that using `to != ""` is not safe as there can be multiple writers, and the layer could potentially be deleted while the operation runs?

Reply:
To maybe overstate the case a bit: either we are keeping a stable API, and this operation, which used to work correctly, needs to continue to work; or we are fine with imposing changes on callers, and in that case we can force the callers to stop using `to != ""`.

It’s probably not _so_ black/white, but I don’t think documenting the problem fundamentally changes the situation. Pragmatically, if there _were_ callers (and I don’t know about any), they wouldn’t notice the warning. Unless, I suppose, we hard-coded a `logrus.Warn*` on every call?

Reply:
Introducing new locks on a code path _with no known users_ seems like a waste and unnecessary risk to me… and I guess runs into my policy of “not reviewing things that add complexity to drivers/overlay”.

I guess this is where @ rhatdan swoops in and merges a PR that outright breaks the API, and I pretend to grumble a bit but I’m secretly relieved 😈 So I suppose we should just do that? At least GitHub shows `ApplyDiffWithDiffer` references only in forks of c/storage and c/image.

Reply:
I'd just break the only caller, but there as well, I am not sure how to keep the CLI :-) working as there is a clear split between creating a layer and populating it

Reply:
Is anyone using the CLI, apart from tests?

Reply:
no

Reply:
… answering myself, CRI-O is now telling users to run the CLI’s `check` command, or rather providing one-off binaries to do so. But I also don’t know about any users using the CLI to trigger partial pulls… and its’ implementation based on a `fileFetcher` is not “partial” unless backed by some hypothetical on-demand-fetch filesystem.

Reply:
`To maybe overstate the case a bit: either we are keeping a stable API, and this operation, which used to work correctly, needs to continue to work; or we are fine with imposing changes on callers, and in that case we can force the callers to stop using to != ""`

Force users to stop using !="". I will merge and @mtrmac can grumble.  Lets move on.

Reply:
I've added a new function `PrepareStagedLayer` that doesn't have the `to` argument.  We can drop the old function once we update c/image

Reply:
Thanks.

Now that the `to != ""` path outright fails, can you follow that to all the consequences, please, so that we don’t leave known-deadlocking code around?
- We won’t need the …`NoLock`/`Finalize` dichotomy (and `Finalize` has no callers)
- The always-`""` parameter of `layerStore.ApplyDiffWithDiffer` can be removed, as well as most of the code
- The always-`""` parameter of `overlay.Driver.ApplyDiffWithDiffer` can be removed, along with users
- (I didn’t realize that the CLI has already been migrated, so that’s simpler.)